### PR TITLE
Adding the ability to plug in custom identity keys and values

### PR DESCRIPTION
**Pull request by bhariharan**
@skyler-b @ashishmo

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java [hunk 1]
```diff
@@ -143,7 +147,41 @@ public UserAccount(String authToken, String refreshToken,
...
+	 * @param communityUrl Community URL.
+	 * @param firstName First Name.
+	 * @param lastName Last Name.
+	 * @param displayName Display Name.
+	 * @param email Email.
+	 * @param photoUrl Photo URL.
+	 * @param thumbnailUrl Thumbnail URL.
+	 * @param customIdentityValues Custom identity values.
+	 */
+	public UserAccount(String authToken, String refreshToken,
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java]: New constructor that takes `customIdentityValues`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java [hunk 2]
```diff
@@ -374,6 +377,21 @@ public UserAccount buildUserAccount(Account account) {
 		if (encThumbnailUrl != null) {
 			thumbnailUrl = SalesforceSDKManager.decryptWithPasscode(encThumbnailUrl, passcodeHash);
 		}
+        Map<String, String> customIdentityValues = null;
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountManager.java]: Encryption/decryption from `AccountManager`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java [hunk 3]
```diff
@@ -581,6 +582,24 @@ public boolean hasNetwork() {
     }
 
     /**
+     * Adds an additional set of custom keys to fetch and store from the identity endpoint.
+     *
+     * @param identityKeys List of custom identity keys.
+     */
+    public void addCustomIdentityKeys(List<String> identityKeys) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java]: New APIs to add custom identity keys.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java [hunk 4]
```diff
@@ -430,6 +448,19 @@ public static synchronized void changePasscode(String oldPass, String newPass) {
                     if (encThumbnailUrl != null) {
                         thumbnailUrl = SalesforceSDKManager.decryptWithPasscode(encThumbnailUrl, oldPass);
                     }
+                    final List<String> customIdKeys = SalesforceSDKManager.getInstance().getCustomIdentityKeys();
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java]: `AccountManager` stuff.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java [hunk 5]
```diff
@@ -198,6 +201,19 @@ public RestClient peekRestClient(Account acc) {
         if (encThumbnailUrl != null) {
             thumbnailUrl = SalesforceSDKManager.decryptWithPasscode(encThumbnailUrl, passcodeHash);
         }
+        final List<String> customIdKeys = SalesforceSDKManager.getInstance().getCustomIdentityKeys();
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java]: `AccountManager` stuff.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java [hunk 6]
```diff
@@ -470,6 +501,14 @@ public static synchronized void changePasscode(String oldPass, String newPass) {
                     if (thumbnailUrl != null) {
                         acctManager.setUserData(account, AuthenticatorService.KEY_THUMBNAIL_URL, SalesforceSDKManager.encryptWithPasscode(thumbnailUrl, newPass));
                     }
+                    if (values != null && !values.isEmpty()) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java]: `AccountManager` stuff.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/util/MapUtil.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    private static final String TAG = "MapUtil";
+
+    /**
+     * Adds a map of key-value pairs extracted from a bundle.
+     *
+     * @param bundle Bundle.
+     * @param keys Keys.
+     * @param map Map to be added to.
+     */
+    public static void addBundleToMap(Bundle bundle, List<String> keys, Map<String, String> map) {
```
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/util/MapUtil.java]: New util class that helps with manipulations from `Bundle` to `Map` to `JSONObject` to support the different types we support.

File: libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/rest/ClientManagerTest.java [hunk 8]
```diff
@@ -105,6 +113,11 @@ public void setUp() throws Exception {
             eq.waitForEvent(EventType.AppCreateComplete, 5000);
         }
         SalesforceSDKManager.getInstance().getPasscodeManager().setPasscodeHash(ClientManagerTest.TEST_PASSCODE_HASH);
+        testIdentityKeys = new ArrayList<>();
```
Comment (bhariharan) [inline libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/rest/ClientManagerTest.java]: Updating tests.

File: libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/rest/RestClientTest.java [hunk 9]
```diff
@@ -120,7 +136,7 @@ public void testGetClientInfo() throws URISyntaxException {
         assertEquals("Wrong email", TEST_EMAIL, restClient.getClientInfo().email);
         assertEquals("Wrong photoUrl", TEST_PHOTO_URL, restClient.getClientInfo().photoUrl);
         assertEquals("Wrong thumbnailUrl", TEST_THUMBNAIL_URL, restClient.getClientInfo().thumbnailUrl);
-
+        assertEquals("Wrong custom identity value", testIdentityValues, restClient.getClientInfo().customIdentityValues);
```
Comment (bhariharan) [inline libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/rest/RestClientTest.java]: Adding test.

**Comment by bhariharan** [comment]
All builds are working, tests are passing, sample apps are working fine.

**Comment by helenren** [comment]
build

**Comment by helenren** [comment]
build

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java (hunk 1 above)
Comment (skyler-b) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java]: Is there any plan to move to a builder pattern for these constructors? Might reduce some of your overhead.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/util/MapUtil.java (hunk 7 above)
Comment (skyler-b) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/util/MapUtil.java]: thanks for being so thorough!

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java (hunk 1 above)
Comment (bhariharan) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccount.java]: @skyler-b I've been meaning to for a while, but haven't found the time to do so. With this change it should allow the ability to plug anything in though.

**Comment by bhariharan** [comment]
All tests are passing, including the new ones I added for this. Sample apps are running fine without the new parameters (tested refresh scenario too). Sample apps are running fine if we pass in special parameters as well (tested refresh scenario too). We should be good to go. @wmathurin Can I merge?

**Pull request merged**